flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/ap/AffinityPropagationBulk.java

```diff
@@ -0,0 +1,449 @@
+		Tuple4<LongValue, LongValue, DoubleValue, DoubleValue>,
+			Tuple3<LongValue, LongValue, DoubleValue>> {
+
+		private Tuple3<LongValue, LongValue, DoubleValue> output =
+			new Tuple3<>(new LongValue(), new LongValue(), new DoubleValue());
+
+		// Receives Tuple6<Trg, MaxValue, MaxNeighbour, SecondMaxValue, is> and Tuple3<src, target, similarity>
+		// and returns a Tuple5<>
+		@Override
+		public Tuple3<LongValue, LongValue, DoubleValue>
+		join(Tuple3<LongValue, LongValue, DoubleValue> similarity,
+			Tuple4<LongValue, LongValue, DoubleValue, DoubleValue> message) {
+
+			output.f0.setValue(similarity.f0.getValue());
+			output.f1.setValue(similarity.f1.getValue());
+			output.f2.setValue(similarity.f2.getValue() + message.f3.getValue());
+
+			return output;
+		}
+	}
+
+	// Create a Tuple4<Trg, MaxValue, MaxNeighbour, SecondMaxValue> reducing the 2 tuples with the max values
+	@ForwardedFields("f1->f0")
+	private static class responsibilityReduceGroup
+		implements GroupReduceFunction<Tuple3<LongValue, LongValue, DoubleValue>,
+		Tuple4<LongValue, DoubleValue, LongValue, DoubleValue>> {
+
+		Tuple4<LongValue, DoubleValue, LongValue, DoubleValue> output = new Tuple4<>(new LongValue(), new DoubleValue(),
+			new LongValue(), new DoubleValue());
+
+		@Override
+		public void reduce(Iterable<Tuple3<LongValue, LongValue, DoubleValue>> maxValues,
+					Collector<Tuple4<LongValue, DoubleValue, LongValue, DoubleValue>> out) throws Exception {
+
+			Long maxNeighbour = Long.valueOf(0);
+			Long trg = Long.valueOf(0);
+			double maxValue = 0;
+			double secondMaxValue = 0;
+
+			for (Tuple3<LongValue, LongValue, DoubleValue> val : maxValues) {
+
+				if(val.f2.getValue() > maxValue){
+					secondMaxValue = maxValue;
+					maxValue = val.f2.getValue();
+					maxNeighbour = val.f0.getValue();
+					trg = val.f1.getValue();
+				}else{
+					secondMaxValue = val.f2.getValue();
+				}
+			}
+
+			output.f0.setValue(trg);
+			output.f1.setValue(maxValue);
+			output.f2.setValue(maxNeighbour);
+			output.f3.setValue(secondMaxValue);
+
+			out.collect(output);
+
+		}
+	}
+
+	// Subtract each responsibility
+	@ForwardedFieldsFirst("f0")
+	@ForwardedFieldsSecond("f0->f1; f1->f0")
+	private static class responsibilityValue
+		implements JoinFunction<Tuple4<LongValue, DoubleValue, LongValue, DoubleValue>,
+		Tuple3<LongValue, LongValue, DoubleValue>,
+		Tuple3<LongValue, LongValue, DoubleValue>> {
+
+		Tuple3<LongValue, LongValue, DoubleValue> output = new Tuple3<>(new LongValue(), new LongValue(),
+			new DoubleValue());
+
+		//Receives Tuple4<Trg, MaxValue, MaxNeighbour, SecondMaxValue> and Tuple3<src, target, similarity>
+		@Override
+		public Tuple3<LongValue, LongValue, DoubleValue>
+		join(Tuple4<LongValue, DoubleValue, LongValue, DoubleValue> maxValues,
+			Tuple3<LongValue, LongValue, DoubleValue> similarity) {
+
+			double responsibility;
+
+			if(similarity.f0.getValue() == maxValues.f2.getValue()){
```

Comment:
This will require keys to be `Comparable` since types cannot be trusted to implement a proper `.equals(Object)`